Add acceleration-based SNS IK base solver with unit test

robodreamer opened the pull request:
This commit adds acceleration-based sns ik base solver including
unit tests. It was written on top of Matt's initial code ported from 
base velocity sns ik solver.

Commit:
Add acceleration-based SNS IK solver with unit test

This commit adds acceleration-based sns IK base solver including
unit tests.

MatthewPeterKelly commented on sns_ik_lib/include/sns_ik/sns_acc_ik_base.hpp:
```diff
@@ -0,0 +1,196 @@
+  typedef std::shared_ptr<SnsAccIkBase> Ptr;
+  typedef std::unique_ptr<SnsAccIkBase> uPtr;
+
+  enum class ExitCode {
```
This duplicated the same enum that is present in the velocity solver. That is fine for now, but ultimately there should be a single base class for both solvers that removes this and other code duplication.

robodreamer replied:
I see. I will note that when I merge them into a single base class.

MatthewPeterKelly commented on sns_ik_lib/src/sns_acc_ik_base.cpp:
```diff
@@ -0,0 +1,408 @@
+ *
+ * @brief The file provides the basic implementation of the SNS-IK acceleration solver
+ *
+ * @author Matthew Kelly
```
Add @author Andy Park to the next line

MatthewPeterKelly commented on sns_ik_lib/include/sns_ik/sns_acc_ik_base.hpp:
```diff
@@ -0,0 +1,196 @@
+ *
+ * @brief The file provides the basic implementation of the SNS-IK acceleration solver
+ *
+ * @author Matthew Kelly
```
Add @author Andy Park to the next line

MatthewPeterKelly commented on sns_ik_lib/src/sns_acc_ik_base.cpp:
```diff
@@ -0,0 +1,408 @@
+  if (!taskScale) { ROS_ERROR("taskScale is nullptr!"); return ExitCode::BadUserInput; }
+  int nTask = ddx.size();
+  if (nTask <= 0) {
+    ROS_ERROR("Bad Input: dx.size() > 0 is required!");
```
dx  -->  ddx

MatthewPeterKelly commented on sns_ik_lib/test/sns_acc_ik_base_test.cpp:
```diff
@@ -0,0 +1,151 @@
+ *  @brief Unit Test: sns_acc_ik_base solver
+ *  @author Matthew Kelly
+ *
+ *    Copyright 2018 Rethink Robotics
```
Add @author Andy Park

MatthewPeterKelly commented on sns_ik_lib/test/sns_acc_ik_base_test.cpp:
```diff
@@ -0,0 +1,151 @@
+
+/*************************************************************************************************/
+
+void checkEqualVector(const Eigen::VectorXd& A, const Eigen::VectorXd& B, double tol)
```
This function and the next one duplicate an existing utility function in the velocity test. These should both be moved to some test utilities header that can be accessed by any test. I suggest making a new file called "test_utilities" that lives next to "rng_utilities".

robodreamer replied:
Good suggestion!

MatthewPeterKelly commented on sns_ik_lib/test/sns_acc_ik_base_test.cpp:
```diff
@@ -0,0 +1,151 @@
+/*************************************************************************************************/
+
+/*
+ * This test is for the solveAccIkBasic()
```
Can you add some more description here?  I think the function name in the existing documentation might also be wrong.

MatthewPeterKelly commented on sns_ik_lib/test/sns_acc_ik_base_test.cpp:
```diff
@@ -0,0 +1,151 @@
+/*************************************************************************************************/
+
+/*
+ * This test is for the solveAccIkBasic()
```
More information in the documentation about the test please.

MatthewPeterKelly reviewed (changes requested):
Looks great!  I made a few suggestions, mostly related to documentation and reducing code duplication.

Commit:
Addressed reviewers' comments

- created test_utilities that include eigen vector data checks
- update author names
- provided more details to testing

MatthewPeterKelly reviewed (approved):
Excellent - ship it!

robodreamer commented:
@IanTheEngineer, hey Ian if you are okay, I will go ahead and merge this pull request.

IanTheEngineer reviewed (approved):
This looks great, Andy! Huge improvement in additional functionality. Thanks!
